morebits.js

```diff
@@ -3647,6 +3664,53 @@ Morebits.wiki.page = function(pageName, status) {
 		return true; // all OK
 	};
 
+	/**
+	 * Determine whether we should provide a watchlist expiry.  Will not
+	 * do so if the page is currently permanently watched, or the current
+	 * expiry is *after* the new, provided expiry.  Only handles strings
+	 * recognized by {@link Morebits.date} or relative timeframes with
+	 * unit it can process.  Relies on the fact that fnCanUseMwUserToken
+	 * requires page loading if a watchlistexpiry is provided, so we are
+	 * ensured of knowing the watch status by the use of this.
+	 *
+	 * @returns {boolean}
+	 */
+	var fnApplyWatchlistExpiry = function() {
+		if (ctx.watchlistExpiry) {
+			if (!ctx.watched) {
+				return true;
+			} else if (typeof ctx.watched === 'string') {
+				var newExp;
+				// Attempt to determine if the new expiry is relative or absolute
+				var rel = ctx.watchlistExpiry.toLowerCase().split(' ');
+				if (rel.length === 2 && !isNaN(parseInt(rel[0], 10)) &&
+					(Morebits.date.unitMap[rel[1]] || Morebits.date.unitMap[rel[1] + 's'])) {
+					// Expiry looks like a relative time (e.g. `1 week`), so treat it like one
+					newExp = new Morebits.date().add(rel[0], rel[1]);
+					if (!newExp.isValid()) {
+						// Oh well!
+						return false;
+					}
+				} else {
+					newExp = new Morebits.date(ctx.watchlistExpiry);
+				}
```

Comment:
```suggestion
				var rel = ctx.watchlistExpiry.toLowerCase().split(' ');
				try {
					newExp = new Morebits.date().add(rel[0], rel[1]);
				} catch() {
					newExp = new Morebits.date(ctx.watchlistExpiry);
				}
```

No need to rely on the implementation details of Morebits.date or duplicate its error handling clauses here. The date#add() will always either return a valid date or throw.

Reply:
I was hoping to avoid unnecessary console messages but yeah, this would be nicer to look at!